### ISSUE TITLE
Swift 6.0.0.4.52 Double deinit called for non-copyable type crossing module boundary

### ISSUE THREAD

**Issue by berikv** [swiftlang/swift#75172]
### Description

A double call to deinit is made for the same instance under very specific conditions:

* Release configuration must be used
* The instance is non-copyable (`~Copyable`)
* The init method must be generic over the argument (`_ scalars: some Collection<Float>`)
* The init method must use the argument (`for _ in zip(0..<15, scalars.indices) {}`
* The init method muts be `@inlinable`
* The call site must be in a separate module
* The call site must call from a generic closure

### Reproduction

```swift
> rm -rf .build; swift build -c release && ./.build/release/clitool
Building for production...
(warnings...)
[7/7] Linking clitool
Build complete! (2.55s)
Init Optional(0x0000600002e18440), refCount 1
Init Optional(0x0000600002e184c0), refCount 1
Deinit Optional(0x0000600002e184c0), refCount 0
Init Optional(0x0000600002e184c0), refCount 1
Deinit Optional(0x0000600002e184c0), refCount 0
Deinit Optional(0x0000600002e18440), refCount 0
Deinit Optional(0x0000600002e18440), refCount -1
DeinitNonCopyableLib/SomeNonCopyable.swift:24: Fatal error: Over freed
[1]    19457 trace trap  ./.build/release/clitool
```

[DeinitNoncopyable.zip](https://github.com/user-attachments/files/16176086/DeinitNoncopyable.zip)

### Expected behavior

No double freeing of non-copyable types

### Environment

swiftc -version
swift-driver version: 1.110 Apple Swift version 6.0 (swiftlang-6.0.0.4.52 clang-1600.0.21.1.3)
Target: arm64-apple-macosx14.0

**Comment by tbkka** [comment]
CC: @kavon @jckarter

**Comment by mrousavy** [comment]
I encountered a similar bug when using `swift::Optional` from C++. I am not sure if this is related to this issue, I'll continue to investigate and maybe create a separate issue report if needed.

**Comment by kavon** [comment]
Thanks for the report! I was able to reproduce this using Xcode 16 Beta 2 / `swiftlang-6.0.0.4.52`.

It appears to have been fixed if you use the `release/6.0` nightly toolchain!

Here's how you can test it to verify:
```
# install the release/6.0 toolchain from https://www.swift.org/install/macos/ 
# Assuming you chose to install for "yourself",
# you'll be able to find it in your home directory under `~/Library/Developer/Toolchains` rather than `/`.

# get the bundle identifier
❯ plutil -extract CFBundleIdentifier raw ~/Library/Developer/Toolchains/swift-6.0-DEVELOPMENT-SNAPSHOT-2024-07-16-a.xctoolchain/Info.plist 
org.swift.600202407161a

# test it using the downloaded toolchain.
❯ rm -rf .build; xcrun --toolchain org.swift.600202407161a swift build -c release && ./.build/release/clitool
Building for production...
[ warnings omitted ]
[7/7] Linking clitool
Build complete! (12.34s)
Init Optional(0x000000014be05cc0), refCount 1
Init Optional(0x000000014be064f0), refCount 1
Deinit Optional(0x000000014be064f0), refCount 0
Init Optional(0x000000014be06350), refCount 1
Deinit Optional(0x000000014be06350), refCount 0
Deinit Optional(0x000000014be05cc0), refCount 0
❯ xcrun --toolchain org.swift.600202407161a swift --version
Apple Swift version 6.0-dev (LLVM febd42d2f7cd87e, Swift c5a97385536617c)
Target: arm64-apple-macosx15.0
```

**Comment by berikv** [comment]
Thank you @kavon, I validated locally using Xcode 16.0 beta 4, and the issue is solved
